presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,274 @@
+            }
+            if (joinDistributionType.equals(FeaturesConfig.JoinDistributionType.REPLICATED) || joinDistributionType.equals(FeaturesConfig.JoinDistributionType.AUTOMATIC)) {
+                possibleJoinNodes.add(joinNode.withDistributionType(REPLICATED));
+                possibleJoinNodes.add(joinNode.flip().withDistributionType(REPLICATED));
+            }
+
+            return planNodeOrdering().min(possibleJoinNodes);
+        }
+
+        private Ordering<PlanNode> planNodeOrdering()
+        {
+            return new Ordering<PlanNode>()
+            {
+                @Override
+                public int compare(PlanNode node1, PlanNode node2)
+                {
+                    PlanNodeCostEstimate node1Cost = lookup.getCumulativeCost(node1, session, symbolAllocator.getTypes());
+                    PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                    return costComparator.compare(session, node1Cost, node2Cost);
+                }
+            };
+        }
+    }
+
+    /**
+     * This method generates all the ways of dividing  totalNodes into two sets
+     * each containing at least one node. It will generate one set for each
+     * possible partitioning. The other partition is implied in the absent values.
+     * In order not to generate the inverse of any set, we always include the 0th
+     * node in our sets.
+     *
+     * @param totalNodes
+     * @return A set of sets each of which defines a partitioning of totalNodes
+     */
+    public static Set<Set<Integer>> generatePartitions(int totalNodes)
+    {
+        Set<Integer> numbers = IntStream.range(0, totalNodes)
+                .boxed()
+                .collect(toImmutableSet());
+        return Sets.powerSet(numbers).stream()
+                .filter(subSet -> subSet.contains(0))
+                .filter(subSet -> subSet.size() < numbers.size())
+                .collect(toImmutableSet());
+    }
+
+    public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph, Set<Integer> leftSources, PlanNodeIdAllocator idAllocator)
+    {
+        List<PlanNode> sources = joinGraph.getSources();
+        ImmutableList.Builder<PlanNode> leftNodes = ImmutableList.builder();
+        Set<Symbol> leftSymbols = new HashSet<>();
+        ImmutableList.Builder<PlanNode> rightNodes = ImmutableList.builder();
+        for (int i = 0; i < sources.size(); i++) {
+            PlanNode source = sources.get(i);
+            if (leftSources.contains(i)) {
+                leftNodes.add(source);
+                leftSymbols.addAll(source.getOutputSymbols());
+            }
+            else {
+                rightNodes.add(source);
+            }
+        }
+
+        InnerJoinPredicateUtils.InnerJoinPushDownResult pushDownResult = sortPredicatesForInnerJoin(leftSymbols, ImmutableList.of(joinGraph.getFilter()), ImmutableList.of());
```

Comment:
> This doesn't work because you do still need to sort the nonInferrable conjuncts for the left and right nodes (you suggest in point 2 that you don't but I don't see how else you'd know what predicates to put there)

Why do you need to do that? To deduce which conjucts should be used in newly created `JoinNode` you just need left and right symbols as in my example (I use `allFilters` and `allInference` as a base). You don't need this information to be stored in `MultiJoinNode`, do you? 
In fact I would rename `MultiJoinNode` to `MultiJoinKey` and just keep `sources` there.